File: frontend/src/types/events.ts
export type ClientEvent =
  | { type: 'audio'; data: string }
  | { type: 'text'; text: string }
  | { type: 'turn_complete' }
  | { type: 'grant_speaking_turn'; agent: string }
  | { type: 'cast_vote'; vote: 'yes' | 'no' | 'abstain' }

export type ServerEvent =
  | { type: 'meeting_status'; payload: { status: string } }
  | { type: 'transcript_update'; payload: { speaker: string; text: string; partial: boolean } }
  | { type: 'audio_chunk'; payload: { data: string; mime: string } }
  | { type: 'turn_complete'; payload: Record<string, never> }
  | { type: 'error'; payload: { message: string } }
  | { type: 'vote_result'; payload: { votes: Record<string, { vote: string }>; result: string } }
  | { type: 'agent_speaking'; payload: { agent: string } }

File: backend/models/events.py
from typing import Literal
from pydantic import BaseModel


class AudioInEvent(BaseModel):
    type: Literal['audio']
    data: str


class TextInEvent(BaseModel):
    type: Literal['text']
    text: str


class TurnCompleteInEvent(BaseModel):
    type: Literal['turn_complete']


class GrantSpeakingTurnInEvent(BaseModel):
    type: Literal['grant_speaking_turn']
    agent: str


class VoteInEvent(BaseModel):
    type: Literal['cast_vote']
    vote: Literal['yes', 'no', 'abstain']


class OutEvent(BaseModel):
    type: str
    payload: dict
